File: crack-bcrypt/golang-bcrypt/bcrypttest/bcrypttest.go
// PT Dymar Jaya Indonesia
// Bcrypt test
// required: go get golang.org/x/crypto/bcrypt
package main

import (
	"flag"
	"fmt"
	"os"

	"golang.org/x/crypto/bcrypt"
)

func HashPassword(password string) (string, error) {
	//cost factor. min:4, max:31
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 10)
	return string(bytes), err
}

func CheckPasswordHash(password, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err == nil
}

func main() {

	flag.Usage = func() {
		fmt.Printf("Usage: %s <data>\n", os.Args[0])
		flag.PrintDefaults()
	}

	flag.Parse()
	if flag.NArg() != 1 {
		flag.Usage()
		os.Exit(1)
	}

	data := os.Args[1]
	hash, _ := HashPassword(data) // ignore error for the sake of simplicity
	
	fmt.Println("data: ",data )
	fmt.Println("prefix:", string(hash)[1:1+2])
	fmt.Println("log rounds: ", string(hash)[4:4+2])
	fmt.Println("salt: ", string(hash)[7:7+22])
	fmt.Println("hash: ", string(hash)[29:])
	fmt.Println("bcrypt: ", hash)

	//match := CheckPasswordHash(password, hash)
	//fmt.Println("Match:   ", match)
}

File: crack-bcrypt/java-bcrypt/bcrypttest.java
// PT Dymar Jaya Indonesia
// Bcrypt test
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;

public class bcrypttest {
    
    public static void main(String[] args) {
	if ((args.length != 1)) {
    System.out.println("Usage: ");
    System.out.println("java bcrypttest <Data>");
    return;
    }
	String data = args[0];
        bcrypttest bcTest = new bcrypttest();
        String bcrypt = bcTest.generateHash(data);
	System.out.println("data: " + data);
	String prefix = bcrypt.substring(1, 1+2);
	System.out.println("prefix: " + prefix);
	String round = bcrypt.substring(4, 4+2);
	System.out.println("log rounds: " + round);
	String salt = bcrypt.substring(7, 7+22);
	System.out.println("salt: " + salt);
	String hashed = bcrypt.substring(29,bcrypt.length());
	System.out.println("hash: " + hashed);
        boolean verify = bcTest.verifyHash(data, bcrypt);
        System.out.println("bcrypt: " +bcrypt);
        //System.out.println("verify: "+verify);
    }
    
    /*
     * generate BCrypt hash
     */
    public String generateHash(String plainText) {
        try {
			String prefix = "$2a"; //version: $2a, $2y, $2b
            int log_rounds = 10; //cost factor. min:4, max:31
            String salt = BCrypt.gensalt(prefix,log_rounds, SecureRandom.getInstance("SHA1PRNG"));
            return BCrypt.hashpw(plainText, salt);
        } catch (NoSuchAlgorithmException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        
        return null;
    }
    
    /*
     * BCrypt check
     */
    public boolean verifyHash(String plainText, String cipher) {
        return BCrypt.checkpw(plainText, cipher);
    }
}
